controllers/spec_preprocessor.go

```diff
@@ -69,47 +74,109 @@ func (r *BackstageReconciler) preprocessSpec(ctx context.Context, backstage bs.B
+
+	// Process SecretFiles
+	if bsSpec.Application.ExtraFiles != nil && bsSpec.Application.ExtraFiles.Secrets != nil {
+		for _, ef := range bsSpec.Application.ExtraFiles.Secrets {
+			secret := &corev1.Secret{}
+			if err := r.addExtConfig(&result, ctx, secret, backstage.Name, ef.Name, ns); err != nil {
+				return result, err
 			}
-			result.ExtraFileConfigMaps[cm.Name] = cm
+			result.ExtraFileSecrets[secret.Name] = *secret
 		}
 	}
 
 	// Process ConfigMapEnvs
 	if bsSpec.Application.ExtraEnvs != nil && bsSpec.Application.ExtraEnvs.ConfigMaps != nil {
 		for _, ee := range bsSpec.Application.ExtraEnvs.ConfigMaps {
-			cm := corev1.ConfigMap{}
-			if err := r.Get(ctx, types.NamespacedName{Name: ee.Name, Namespace: ns}, &cm); err != nil {
-				return result, fmt.Errorf("failed to get configMap %s: %w", ee.Name, err)
+			cm := &corev1.ConfigMap{}
+			if err := r.addExtConfig(&result, ctx, cm, backstage.Name, ee.Name, ns); err != nil {
+				return result, err
+			}
+			result.ExtraEnvConfigMaps[cm.Name] = *cm
+		}
+	}
+
+	// Process SecretEnvs
+	if bsSpec.Application.ExtraEnvs != nil && bsSpec.Application.ExtraEnvs.Secrets != nil {
+		for _, ee := range bsSpec.Application.ExtraEnvs.Secrets {
+			secret := &corev1.Secret{}
+			if err := r.addExtConfig(&result, ctx, secret, backstage.Name, ee.Name, ns); err != nil {
+				return result, err
 			}
-			result.ExtraEnvConfigMaps[cm.Name] = cm
+			result.ExtraEnvSecrets[secret.Name] = *secret
 		}
 	}
 
 	// Process DynamicPlugins
 	if bsSpec.Application.DynamicPluginsConfigMapName != "" {
-		cm := corev1.ConfigMap{}
-		if err := r.Get(ctx, types.NamespacedName{Name: bsSpec.Application.DynamicPluginsConfigMapName,
-			Namespace: ns}, &cm); err != nil {
-			return result, fmt.Errorf("failed to get ConfigMap %v: %w", cm, err)
+		cm := &corev1.ConfigMap{}
+		if err := r.addExtConfig(&result, ctx, cm, backstage.Name, bsSpec.Application.DynamicPluginsConfigMapName, ns); err != nil {
+			return result, err
 		}
-		result.DynamicPlugins = cm
+		result.DynamicPlugins = *cm
 	}
 
 	return result, nil
 }
+
+func (r *BackstageReconciler) addExtConfig(config *model.ExternalConfig, ctx context.Context, obj client.Object, backstageName, objectName, ns string) error {
+
+	lg := log.FromContext(ctx)
+
+	if err := r.Get(ctx, types.NamespacedName{Name: objectName, Namespace: ns}, obj); err != nil {
+		if _, ok := obj.(*corev1.Secret); ok && errors.IsForbidden(err) {
+			return fmt.Errorf("warning: Secrets GET is forbidden, updating Secrets may not cause Pod recreating")
+		}
+		return fmt.Errorf("failed to get external config from %s: %s", objectName, err)
+	}
+
+	if err := config.AddToSyncedConfig(obj); err != nil {
+		return fmt.Errorf("failed to add to synced %s: %s", obj.GetName(), err)
+	}
+
+	if obj.GetLabels() == nil {
+		obj.SetLabels(map[string]string{})
+	}
+	if obj.GetAnnotations() == nil {
+		obj.SetAnnotations(map[string]string{})
+	}
+
+	autoSync := true
+	autoSyncStr, ok := os.LookupEnv(AutoSyncEnvVar)
```

Comment:
Also, could you document this (from a user standpoint, I mean)?

Reply:
yes, on CSV for OLM (like custom images), directly on manifest otherwise (like fsGroup)

As commented, documentation will come later (with configuration docs)